infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/utils/LogReplicationConfigManager.java

```diff
@@ -67,49 +87,183 @@ public class LogReplicationConfigManager {
 
     private Table<VersionString, Version, Uuid> pluginVersionTable;
 
+    // In memory entries of registry table, which will be refreshed when config is syncing with registry table.
+    // This is needed to avoid inconsistency between config fields (for example, after streamsToReplicate is fetched,
+    // some tables opened and confirmedNoisyStreams will be read with potential conflict info)
+    private List<Map.Entry<TableName, CorfuRecord<TableDescriptors, TableMetadata>>> registryTableEntries;
+
+    private long lastRegistryTableLogTail;
+
+    private long lastRegistryTableTrimMark;
+
     /**
-     * Used for testing purpose only.
+     * Used for non-upgrade testing purpose only. Note that this constructor will keep the version table in
+     * uninitialized state, in which case LR will be constantly considered to be not upgraded.
      */
     @VisibleForTesting
     public LogReplicationConfigManager(CorfuRuntime runtime) {
+        this.rt = runtime;
         this.corfuStore = new CorfuStore(runtime);
         this.pluginConfigFilePath = EMPTY_STR;
+        this.lastRegistryTableLogTail = Address.NON_ADDRESS;
+        this.lastRegistryTableTrimMark = Address.NON_ADDRESS;
     }
 
     public LogReplicationConfigManager(CorfuRuntime runtime, String pluginConfigFilePath) {
+        this.rt = runtime;
         this.pluginConfigFilePath = pluginConfigFilePath;
         this.corfuStore = new CorfuStore(runtime);
+        this.lastRegistryTableLogTail = Address.NON_ADDRESS;
+        this.lastRegistryTableTrimMark = Address.NON_ADDRESS;
+        initLogReplicationVersionPlugin();
+        setupVersionTable();
+    }
+
+    private void initLogReplicationVersionPlugin() {
+        log.info("Plugin :: {}", pluginConfigFilePath);
+        LogReplicationPluginConfig config = new LogReplicationPluginConfig(pluginConfigFilePath);
```

Comment:
It seems like this class should be only instantiated once and passed, we can revisit this later.

Reply:
I think it's only instantiated here?

Reply:
In 5 other places in fact.